[FEATURE ember-runtime-enumerable-includes] Implements Array.includes and deprecates Array.contains

bmeurant opened the pull request:
Implementation of emberjs/rfcs#136

TODO : 
- [x]  Implement `Enumerable.includes` 
- [x]  Implement `Array.includes`
- [x]  Deprecate `contains` under feature flag
- [x]  Replace usages of `contains` by `includes`
- [x]  Update docs, guides (emberjs/guides#1439) & API. To be merged once feature enabled.
- [x] Write deprecation guide (emberjs/website#2600) & add url in deprecate call
- [x] Rebase / squash

bmeurant committed:
[FEATURE ember-runtime-enumerable-includes] Introduce Enumerable#includes
and deprecate Enumerable#contains

alexspeller commented:
Oh awesome, thanks for getting this implementation done!

bmeurant commented:
Here is a first proposal for emberjs/rfcs#136. Work is in progress and feedbacks expected.

I made the asumption that we have to keep `Enumerable` & `Array` APIs consistent and then have to provide a consistent `includes` method in `Enumerable` (with no index). Let me know if this is not expected.

Aliasing `contains` with the new `includes` method can break some existing apps because, according to the spec (https://tc39.github.io/ecma262/2016/#sec-array.prototype.includes), `includes` relies on the `SameValueZero` algorithm and actual `contains` relies on strict equality leading to differences in `NaN` detections (I will document it). Can we still make this ?

mixonic commented:
@bmeurant Thanks :-)

`contains` should not be aliased to `includes`. Contains should keep its current implementation, and we can deprecate it though we will likely do that in a separate PR. In this PR it should simply retain its current behavior.

bmeurant commented:
@mixonic thx for your feedback.

I updated this PR to keep current implementation for `Enumerable.contains` & `Array.contains`.

> we can deprecate it though we will likely do that in a separate PR

Does it means that I should also remove `@deprecated` and deprecation warning for now ?

rwjblue commented:
I believe that the deprecation should be included here when the feature flag is enabled.

mixonic commented:
@rwjblue I am easily convinced, was just erring on the conservative side. @bmeurant please include the deprecation behind the feature flag! :-)

bmeurant commented:
@mixonic I added feature flag and move deprecation behind it. Let me know if I forgotten something.

I also removed `@deprecated` from API because `contains` is not actuelly deprecated if feature is not enabled, right ?

But I was wondering: doing that, users will loose the information that `contains` will be deprecated soon and that they should use `includes` instead. Is it exists a way to provide this kind of information in APIs ?

mixonic commented:
@bmeurant Yeah the documentation may require some followup tweaks. Feature flagging things like the `@deprecated` flag is tricky :-/. I think you should feel free to create a followup issue of things to do once the feature is "go'd" for beta.

mixonic commented on FEATURES.md:
```diff
@@ -56,3 +56,8 @@ for a detailed explanation.
+
+* `ember-runtime-enumerable-contains`
+
+Deprecates `Enumerable.contains` and `Array.contains` in favor of `Enumerable.includes` and `Array.includes` 
```
Again here I lean toward `Enumerable#contains`, per https://github.com/emberjs/website/pull/2600#discussion_r64891064. It is a Rubyish approach, curious if there are other thoughts.

alexspeller replied:
I think `Enumerable#contains` is the convention in Ember even though it was originally a rubyism - at least it has the advantage that it's not valid syntax for something that is completely different

mixonic commented on packages/ember-runtime/lib/mixins/enumerable.js:
```diff
@@ -239,6 +248,27 @@ var Enumerable = Mixin.create({
+    @return {Boolean} `true` if object is found in the enumerable.
+    @public
+  */
+  includes(obj) {
```
Can you add an assertion that there is no second argument? Something like:

```js
  includes(obj) {
    assert('Enumerable#includes cannot accept a second argument "startAt" as enumerable items are unordered.', arguments.length === 1);
```

mixonic commented on packages/ember-runtime/lib/mixins/mutable_array.js:
```diff
@@ -388,7 +388,7 @@ export default Mixin.create(EmberArray, MutableEnumerable, {
   */
   addObject(obj) {
-    if (!this.contains(obj)) {
+    if (!this.includes(obj)) {
```
This is a small breaking change. I've left a note that it be called out in the deprecation guide https://github.com/emberjs/website/pull/2600#issuecomment-222124560. Here and `without`.

bmeurant replied:
Sure. Moreover, I think I have to use includes instead of contains only behind feature flag, right?

mixonic replied:
👍 yup

mixonic commented on packages/ember-runtime/tests/suites/array/includes.js:
```diff
@@ -0,0 +1,35 @@
+
+var suite = SuiteModuleBuilder.create();
+
+suite.module('includes');
```
Should include a test/tests without a starting position

bmeurant replied:
Tests without starting position are imported from Enumerable test suite, no ?

mixonic replied:
I'm unsure what you mean, maybe I am missing something. This code tests the implementation of `Array#includes`. How are the tests for `Enumerable#includes` applied to an array?

mixonic commented:
@bmeurant heh, there are no tests for `Array#contains`?! Can you add one asserting basic behavior and the deprecation please?

This is looking pretty good after some tweaks!

mixonic commented on packages/ember-runtime/lib/mixins/array.js:
```diff
@@ -214,9 +216,63 @@ export default Mixin.create(Enumerable, {
+    @return {Boolean} `true` if object is found in the array.
+    @public
+  */
+  includes(obj, startAt) {
```
This function (and the enumerable one) must be added only when the feature flag is enabled. For example:

```js
let ArrayMixin = Mixin.create(Enumerable, {
 /* snip */
});
if (isEnabled('ember-runtime-enumerable-includes')) {
  ArrayMixin.reopen({
    includes(obj, startAt) {
      /* ... */
    }
  });
}
export default ArrayMixin;
```

mixonic commented on packages/ember-runtime/tests/suites/enumerable/includes.js:
```diff
@@ -0,0 +1,28 @@
+
+suite.module('includes');
+
+suite.test('includes returns true if items is in enumerable', function() {
```
Once the method itself is behind a feature flag, you will also need to wrap these tests inside an `isEnabled` check.

mixonic commented on features.json:
```diff
@@ -10,6 +10,7 @@
     "ember-runtime-computed-uniq-by": null,
-    "ember-improved-instrumentation": null
+    "ember-improved-instrumentation": null,
+    "ember-runtime-enumerable-contains": null
```
Lastly the feature flag here seems poorly named. Perhaps `ember-runtime-enumerable-includes`

mixonic commented on packages/ember-runtime/lib/mixins/enumerable.js:
```diff
@@ -239,6 +248,27 @@ var Enumerable = Mixin.create({
+      return item === obj || (item !== item && obj !== obj);
+    });
+
+    return found !== undefined;
```
Wouldn't this break for `includes(undefined)`? Can you add a test?

bmeurant replied:
Right ! I missed it, sorry :-(

bmeurant commented:
Thx for feedbacks, I'll work on it ASAP

bmeurant commented:
It seemed to me that Array#contains tests are also imported and run from Enumerable#contains tests. Did I missed something ?

mixonic commented:
Ah I see @bmeurant :-) right there is no `contains` on arrays, only on enumerable 👍

bmeurant commented on packages/ember-runtime/lib/mixins/mutable_array.js:
```diff
@@ -388,7 +389,9 @@ export default Mixin.create(EmberArray, MutableEnumerable, {
   */
   addObject(obj) {
-    if (!this.contains(obj)) {
+    var included = isEnabled('ember-runtime-enumerable-includes') ? this.includes(obj) : this.contains(obj);
```
Impacts on `addObject` are not so small because using `includes` instead of  `contains` here will allow to add `null` and `undefined` values into an Enumerable ... Is it expected ?

mixonic replied:
Unfortunately the ternary syntax is not supported by feature flag stripping. Please rewrite this to use an `if` statement.

bmeurant commented:
I wanted to add a test to verify assertion calling `Enumerable#includes` with a second argument but, because tests defined on Enumerable are also run on Array suites, I did not find a good way to do it. Any advice ?

bmeurant commented on packages/ember-runtime/lib/mixins/enumerable.js:
```diff
@@ -811,7 +820,9 @@ var Enumerable = Mixin.create({
     @public
   */
   without(value) {
```
I think we have a problem line 832 during `k !== value`. If the feature is enabled, `without` become inconsistent because `NaN !== NaN` is `true`

bmeurant commented on packages/ember-runtime/lib/mixins/enumerable.js:
```diff
@@ -1118,4 +1127,63 @@ if (isEnabled('ember-runtime-computed-uniq-by')) {
+      return found;
+    },
+
+    without(value) {
```
I prefered replacing `without` behind feature flag to not add multiple `isEnabled` calls in original method

mixonic replied:
👍

bmeurant commented on packages/ember-runtime/lib/mixins/array.js:
```diff
@@ -214,6 +216,14 @@ export default Mixin.create(Enumerable, {
+      deprecate(
+        '`Enumerable#contains` is deprecated, use `Enumerable#includes` instead.',
+        false,
+        { id: 'ember-runtime.enumerable-contains', until: '3.0.0', url: 'http://emberjs.com/deprecations/v2.x#toc_enumerable-contains' }
```
I did not change deprecation id because feature is about `includes` but deprecation is about  `contains`.

But I don't know if it is accepted. Let me know

mixonic replied:
the deprecation id and feature id are totally distinct. Seems fine to use different names 👍

mixonic commented on packages/ember-runtime/lib/mixins/enumerable.js:
```diff
@@ -1118,4 +1127,63 @@ if (isEnabled('ember-runtime-computed-uniq-by')) {
+    },
+
+    without(value) {
+      if (!this.includes(value)) {
```
Calling `includes` before the `forEach` means we walk the contents of the array twice. Additionally it seems a bit unexpected for `without()` to sometimes return a new array, and sometimes return the array passed in.

I'd prefer to drop the `includes` check and just run the array builder.

bmeurant replied:
I think I'd also prefer but it seems that returning the same instance if not included (or not contained), is part of the API since it is tested here : https://github.com/emberjs/ember.js/blob/master/packages/ember-runtime/tests/suites/enumerable/without.js#L19

Does this PR should also change this behaviour?

I could make it, of course, and update the deprecation guide. But in any case, I propose to update `without` API doc to explicitely mention it.

mixonic replied:
👍  to updating the docs and keeping the current behavior. Great catch.

The `without` method on enumerables and on mixins just should just 🔥 🔥 🔥. On enumerables we can just use `reject` which seems more sane, and on mixins the functionality is only used for native array prototype application. I attempted killing it once and ran out of steam.

mixonic commented on packages/ember/tests/component_registration_test.js:
```diff
@@ -39,9 +40,13 @@ function cleanup() {
   keys(helpers).
     forEach((name) => {
-      if (!originalHelpers.contains(name)) {
+      included = isEnabled('ember-runtime-enumerable-includes') ? originalHelpers.includes(name) : originalHelpers.contains(name);
```
Unfortunately the ternary syntax is not supported by feature flag stripping. Please rewrite this to use an `if` statement.

mixonic commented:
@bmeurant it seems ok to add a test for `contains` without an offset in `packages/ember-runtime/tests/mixins/enumerable_test.js`, since it only impacts enumerables.

mixonic commented:
Thanks for your continued work on this 😄 it looks really close, and quite nice!

mixonic commented:
Looking nice. A rebase/squash to some number is semantically meaningful commits (one or more) would be great, each prefixed with `[FEATURE ember-runtime-enumerable-includes]` would be great.

bmeurant commented:
Done :ok_hand:

homu commented:
:umbrella: The latest upstream changes (presumably #13328) made this pull request unmergeable. Please resolve the merge conflicts.

mixonic commented:
Looks pretty great to me @bmeurant! Thank you! I'll get someone else on core to review and we can :shipit: with their 👍

rwjblue commented on packages/ember-runtime/lib/mixins/array.js:
```diff
@@ -573,3 +583,55 @@ export default Mixin.create(Enumerable, {
+      var len = get(this, 'length');
+      var idx, currentObj;
+
+      if (Ember.isNone(startAt)) {
```
Should not use the global here (`Ember.isNone`), if we need `isNone` it should be imported.  Can we just do the `null` / `undefined` check ourselves?

bmeurant replied:
@rwjblue using the global here was surely a mistake, sorry. `isNone` is already imported and used in `slice`. I'll update. But why do you prefer to not use imported `isNone` and doing `null` / `undefined` check ourselves in this case?

bmeurant replied:
Anyway, explicit null check is unnecessary as we only iterating on `idx = startAt`. I removed it.

rwjblue commented:
This looks good, only one small tweak and this is good to go...

mixonic commented:
This is great! Thank you @bmeurant!

rwjblue commented:
FYI - I created https://github.com/rwjblue/ember-runtime-enumerable-includes-polyfill to make this a bit easier for addons to avoid deprecations, use the newer syntax, and continue to support older versions of Ember.
